### PR TITLE
Alerting: Handle namespace and group query string params in Ruler API

### LINKED ISSUE #84481: Alerting: Handle slashes in rules API endpoint
Prometheus and Mimir rules containing slashes in namespaces and group names cause 404 errors due to slash character being encoded by middleware proxies before reaching grafana server

There are 2 possible solutions to this problem:

1. Pass parameters that currently go in the path to query parameters (like /rules?group=foo%2Fbar&namespace=with%2Fslashes)
2. Pass them in request body of a POST request 

1) we have paths like `/ruler/{DatasourceUID}/api/v1/rules/{Namespace}/{Groupname}` in the case of slashes in folder name\group we do `/ruler/{DatasourceUID}/api/v1/rules/QUERY_NAMESPACE/QUERY_GROUP?namespace=with%2Fslashes&group=with%2Fslashes`. 
This way we do not need to modify any routing, and just update the proxy URL builder. 
Pros: the path can work as normal and do not break anyone who relies upon those APIs in OSS or GE.

2) will require changes in endpoints and models. A new set of endpoints would make sense. The model for those endpoints could be a wrapper for an old model with group name and namespace.

**Related bug reports**
https://github.com/grafana/grafana/issues/83088

### PR DESCRIPTION
**Pull request by konrad147**
**What is this feature?**
This PR implements a workaround for handling slashes in namespaces and groups names. These are used to build API urls, however some proxies might decode the path and break the initial url structure resulting in 404 errors

Fixes #84481

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by yuri-tceretian — COMMENTED** [review]
Overall, LGTM. Two things that I would recommend:
1. Use constants instead of literals everywhere where it's possible
2. Check if query param is empty and return error if it's true.

File: pkg/services/ngalert/api/lotex_ruler.go [hunk 1]
```diff
@@ -177,7 +217,13 @@ func (r *LotexRuler) RoutePostNameRulesConfig(ctx *contextmodel.ReqContext, conf
...
 		return ErrResp(500, err, "Failed marshal rule group")
 	}
-	u := withPath(*ctx.Req.URL, fmt.Sprintf("%s/%s", legacyRulerPrefix, ns))
+
+	var finalNamespace = ns
+	if ns == "QUERY_NAMESPACE" {
+		finalNamespace = ctx.Query("namespace")
+	}
+
+	u := withPath(*ctx.Req.URL, fmt.Sprintf("%s/%s", legacyRulerPrefix, finalNamespace))
```
Comment (yuri-tceretian) [inline pkg/services/ngalert/api/lotex_ruler.go:226]:
```suggestion
	u := withPath(*ctx.Req.URL, fmt.Sprintf("%s/%s", legacyRulerPrefix, url.PathEscape(finalNamespace)))
```

File: public/app/features/alerting/unified/api/ruler.ts [hunk 2]
```diff
@@ -20,34 +20,96 @@ export interface RulerRequestUrl {
...
+  if (isGrafanaDatasource) {
+    return {
+      namespace: (namespace: string) => ({ namespace, searchParams: {} }),
+      group: (group: string) => ({ group, searchParams: {} }),
+    };
+  }
+
+  return {
+    namespace: (namespace: string) => {
+      const containsSlash = namespace.includes('/');
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/api/ruler.ts:92]: Can we use `checkForPathSeparator()` instead? it should also check for `\`;

File: public/app/features/alerting/unified/api/ruler.ts [hunk 3]
```diff
@@ -20,34 +20,96 @@ export interface RulerRequestUrl {
...
+  group: (group: string) => GroupUrlParams;
+}
+
+function getQueryDetailsProvider(rulerConfig: RulerDataSourceConfig): RulerQueryDetailsProvider {
+  const isGrafanaDatasource = rulerConfig.dataSourceName === GRAFANA_RULES_SOURCE_NAME;
+
+  const externalRulerSearchParams = {
+    subtype: rulerConfig.apiVersion === 'legacy' ? 'cortex' : 'mimir',
+  };
+
```
Comment (konrad147) [inline public/app/features/alerting/unified/api/ruler.ts:81]: We can skip adding `subtype` here. The ruler builder should handle that

File: public/app/features/alerting/unified/api/ruler.ts [hunk 4]
```diff
@@ -20,34 +20,96 @@ export interface RulerRequestUrl {
...
+  // For Grafana we only proxy namespace and group parameters. No need to deal with slashes
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/api/ruler.ts:82]: We should for `group` :)

**Comment by gillesdemey** [comment]
We should probably also remove the artificial limitations we have now with regards to the path separator. Do you intend to remove the `checkForPathSeparator` validations in this PR?

File: pkg/services/ngalert/api/lotex_ruler.go (hunk 1 above)
Comment (gillesdemey) [reply pkg/services/ngalert/api/lotex_ruler.go]: @konrad147 I think we still need to escape the input here?

**Comment by konrad147** [comment]
> We should probably also remove the artificial limitations we have now with regards to the path separator. Do you intend to remove the `checkForPathSeparator` validations in this PR?

I'd say no. This implementation is only a workaround.

**Review by gillesdemey — APPROVED** [review]
FE changes LGTM!

**Review by yuri-tceretian — APPROVED** [review]
LGTM! Great job!

File: pkg/services/ngalert/api/lotex_ruler.go (hunk 1 above)
Comment (konrad147) [reply pkg/services/ngalert/api/lotex_ruler.go]: It's done

**Comment by diegommm** [comment]
Hi! Greetings from the Release Guild. This is a friendly reminder that the Steady [11.2.x Release Branch](https://github.com/grafana/grafana/tree/v11.2.x) was cut this week, and that it includes the changes committed up to https://github.com/grafana/grafana/commit/b4126d3bce3627e9460e2db0e870c8d984be8e84. This PR was merged after it, so we wanted to let you know that you can add the `backport v11.2.x` label to to backport it and make it part of the next 11.2.x Patch Release, just in case you may have missed the news.

Thank you!

**Pull request merged**